test/async-observer.test.js

```diff
@@ -324,6 +324,86 @@ describe('async observer', function() {
       }
     );
   });
+
+  it('should call after operation hook on error', function(done) {
+    const context = {
+      req: {},
+    };
+    const operationError = new Error('The operation failed without result');
+    let passesBy = 0;
+
+    function fn(context, done) {
+      process.nextTick(() => {
+        done(operationError);
+      });
+    }
+
+    TestModel.observe('after execute error', function(ctx, next) {
+      passesBy++;
+      next();
+    });
+
+    TestModel.notifyObserversAround('execute', context, fn, (err, ctx) => {
+      passesBy.should.eql(1);
+      err.message.should.eql(operationError.message);
+      ctx.error.message.should.eql(operationError.message);
+      done();
+    });
+  });
+
+  it('should call after operation hook on error while overwriting error', function(done) {
+    const context = {
+      req: {},
+    };
+    const operationError = new Error('The operation failed without result');
+    const overwriteError = new Error('Overwriting the original error');
+    let passesBy = 0;
+
+    function fn(context, done) {
+      process.nextTick(() => {
+        done(operationError);
+      });
+    }
+
+    TestModel.observe('after execute error', function(ctx, next) {
+      passesBy++;
+      next(overwriteError);
+    });
+
+    TestModel.notifyObserversAround('execute', context, fn, (err, ctx) => {
+      passesBy.should.eql(1);
+      err.message.should.eql(overwriteError.message);
+      ctx.error.message.should.eql(operationError.message);
+      done();
+    });
+  });
+
+  it('should call after operation hook on error while allowing to change err', function(done) {
+    const context = {
+      req: {},
+    };
+    const operationError = new Error('The operation failed without result');
+    let passesBy = 0;
+
+    function fn(context, done) {
+      process.nextTick(() => {
+        done(operationError);
+      });
+    }
+
+    TestModel.observe('after execute error', function(ctx, next) {
+      passesBy++;
+      const err = ctx.error;
+      ctx.end(err, ctx);
```

Comment:
I don't remember what is `ctx.end` doing, could you please help me to understand what's the expected outcome of this call?

Also AFAICT, you are not changing the error, you keep using the same `operationError` instance. What am I missing?

Reply:
From https://loopback.io/doc/en/lb3/Connector-hooks.html#before-execute 
> To terminate the invocation, call ctx.end(err, result)
Imho it should early terminate the request with a custom response.

Reply:
I see, thank you for the link. The documentation is for "before execute" hook where it's possible to skip execution of the actual command. In "after execute" and "after execute error" hooks, the command was already executed, thus I don't see what would be the point of calling `ctx.end`?

Anyhow, I don't mind to test this scenario, but please rework the test to make it more clear what is the use case being tested.

Reply:
IMO, you should call `next` instead of `ctx.end` to signal the error.

```suggestion
      next(err);
```